src/HtmlProcessor/CssVariableEvaluator.php

```diff
@@ -0,0 +1,197 @@
+                return $this->replaceVariablesInPropertyValue($fallbackValue);
+            } else {
+                return $matches[0];
+            }
+        }
+    }
+
+    /**
+     * Regular expression based on {@see https://stackoverflow.com/a/54143883/2511031 a StackOverflow answer}.
+     */
+    private function replaceVariablesInPropertyValue(string $propertyValue): string
+    {
+        return (new Preg())->replaceCallback(
+            '/
+                var\\(
+                    \\s*+
+                    # capture variable name including `--` prefix
+                    (
+                        --[^\\s\\),]++
+                    )
+                    \\s*+
+                    # capture optional fallback value
+                    (?:
+                        # capture separator to confirm there is a fallback value
+                        (,)\\s*
+                        # begin capture with named group that can be used recursively
+                        (?<recursable>
+                            # begin named group to match sequence without parentheses, except in strings
+                            (?<noparentheses>
+                                # repeated zero or more times:
+                                (?:
+                                    # sequence without parentheses or quotes
+                                    [^\\(\\)\'"]++
+                                    |
+                                    # string in double quotes
+                                    "(?>[^"\\\\]++|\\\\.)*"
+                                    |
+                                    # string in single quotes
+                                    \'(?>[^\'\\\\]++|\\\\.)*\'
+                                )*+
+                            )
+                            # repeated zero or more times:
+                            (?:
+                                # sequence in parentheses
+                                \\(
+                                    # using the named recursable pattern
+                                    (?&recursable)
+                                \\)
+                                # sequence without parentheses, except in strings
+                                (?&noparentheses)
+                            )*+
+                        )
+                    )?+
+                \\)
+            /x',
+            \Closure::fromCallable([$this, 'getPropertyValueReplacement']),
+            $propertyValue
+        );
+    }
+
+    /**
+     * @param array<non-empty-string, string> $declarations
+     *
+     * @return array<non-empty-string, string>|false `false` is returned if no substitutions were made.
+     */
+    private function replaceVariablesInDeclarations(array $declarations)
+    {
+        $substitutionsMade = false;
+        $result = \array_map(
+            function (string $propertyValue) use (&$substitutionsMade): string {
```

Comment:
Using all these used array functions feels to me like we maybe could use another class structure to have more type safety and to make the code more communicative. What do you think?

Reply:
Not sure what you mean or are suggesting.  The method performs a transformation on each element of an array, for which `array_map` does the job.  The reason to keep track of whether this transformation made any changes is an optimization.  PHP's copy-on-write handling of arrays helps, but if a change is made, comparing `$newArray !== $oldArray` means checking each element (until the difference is found).  (If the array is unmodified, the test is instant, since internally they'll still be referencing the same array.)